src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardNativeWrapper.cs

```diff
@@ -0,0 +1,292 @@
+            }
+
+            using (var operationContext = InitializeOperationContext(compressedBufferInfo))
+            {
+                yield return ProcessCompressedOutput(operationContext, (zcs, buffer) => Zstandard64NativeMethods.ZSTD_endStream(zcs, buffer));
+            }
+
+            int ProcessCompressedOutput(OperationContext context, Action<IntPtr, NativeBufferInfo> outputAction)
+            {
+                ConfigureNativeBuffer(
+                    _outputNativeBuffer,
+                    context.CompressedPinnedBuffer,
+                    _recommendedZstreamOutputSize);
+
+                outputAction(_zstreamPointer, _outputNativeBuffer);
+
+                return (int)_outputNativeBuffer.Position;
+            }
+        }
+
+        // private methods
+        private void ConfigureNativeBuffer(NativeBufferInfo buffer, PinnedBuffer pinnedBuffer, ulong size)
+        {
+            buffer.DataPointer = pinnedBuffer?.IntPtr ?? IntPtr.Zero;
+            buffer.Size = size;
+            buffer.Position = 0;
+        }
+
+        private void InitializeIfNotAlreadyInitialized()
+        {
+            if (!_operationInitialized)
+            {
+                _operationInitialized = true;
+
+                switch (_compressionMode)
+                {
+                    case CompressionMode.Compress:
+                        Zstandard64NativeMethods.ZSTD_initCStream(_zstreamPointer, _compressionLevel); // start a new compression operation
+                        break;
+                    case CompressionMode.Decompress:
+                        Zstandard64NativeMethods.ZSTD_initDStream(_zstreamPointer); // start a new decompression operation
+                        break;
+                }
+            }
+        }
+
+        // nested types
+        internal class OperationContext : IDisposable
+        {
+            private readonly PinnedBuffer _uncompressedPinnedBuffer;
+            private readonly PinnedBuffer _compressedPinnedBuffer;
+
+            public OperationContext(PinnedBuffer uncompressedPinnedBuffer, PinnedBuffer compressedPinnedBuffer)
+            {
+                _uncompressedPinnedBuffer = uncompressedPinnedBuffer; // can be null
+                _compressedPinnedBuffer = Ensure.IsNotNull(compressedPinnedBuffer, nameof(compressedPinnedBuffer));
+            }
+
+            public PinnedBuffer UncompressedPinnedBuffer => _uncompressedPinnedBuffer; // external data
+            public PinnedBuffer CompressedPinnedBuffer => _compressedPinnedBuffer; // internal data
+
+            // public methods
+            public void Dispose()
+            {
+                _uncompressedPinnedBuffer?.Dispose(); // PinnedBuffer.Dispose suppresses all errors
+                _compressedPinnedBuffer.Dispose();
+            }
+        }
+    }
+
+    internal class BufferInfo
```

Comment:
There is a standard .NET class called `ArraySegment<byte>` which could have been used here.

But... since this is vendored code we might as well leave it as is.